### PR TITLE
Fix 'set' skip index not working with IN, indexHint(), and subqueries

### PR DESCRIPTION
**Pull request by al13n321**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed 'set' skip index not working with IN and indexHint()

Example:
```
create table a (k UInt64, v UInt64, index i (v) type set(100) granularity 2) engine MergeTree order by k;
insert into a select number, intDiv(number, 4096) from numbers(1000000);
select sum(1+ignore(*)) from a where indexHint(v in (20, 40));
```

Without the fix it outputs `1000000` (i.e. reads all 1M rows), with the fix `32768` (i.e. reads 4 granules, as expected).

Also works with subqueries now:
```
select sum(1+ignore(*)) from a where indexHint(v in (select 20 union all select 40));
```

* The IN thing got broken by https://github.com/ClickHouse/ClickHouse/pull/58657 - it (inadvertently?) disabled the index if the expression contains a constant set.
* `indexHint()` got broken some time before that.
* Subqueries were probably never supported.

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit fed5bed14212f08374a6c06846c33cafb19f2049 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62083/fed5bed14212f08374a6c06846c33cafb19f2049/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Storages/MergeTree/MergeTreeIndexSet.cpp [hunk 1]
```diff
@@ -326,6 +332,26 @@ bool MergeTreeIndexConditionSet::mayBeTrueOnGranule(MergeTreeIndexGranulePtr idx
...
+    const auto & adaptor = typeid_cast<const FunctionToFunctionBaseAdaptor &>(*node.function_base);
+    const auto & index_hint = typeid_cast<const FunctionIndexHint &>(*adaptor.getFunction());
+    if (!result_dag_or_null)
+        return index_hint.getActions()->getOutputs();
+
+    /// Import the DAG and map argument pointers.
+    ActionsDAGPtr actions_clone = index_hint.getActions()->clone();
+    chassert(storage);
+    *storage = actions_clone->getOutputs();
+    result_dag_or_null->mergeNodes(std::move(*actions_clone));
```
Comment (jkartseva) [inline src/Storages/MergeTree/MergeTreeIndexSet.cpp:351]: Will the node pointers copied to `store` remain valid after actions are passed to `mergeNodes`?

**Review by jkartseva — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@al13n321, fuzzers have found a lot of errors.

**Comment by al13n321** [comment]
The failed test is unrelated (`SELECT * FROM loop('', '')` crashes the server - candidate for the shortest query that does that :) ).

**Comment by DerekChia** [comment]
@al13n321 possible to backport at least to 24.3 LTS? It's been broken for a while - https://github.com/ClickHouse/ClickHouse/issues/65141

**Comment by clickgapai** [comment]
Hi @al13n321 @jkartseva — while reviewing this PR I found the following:

- **Coverage gap:** `set` skip index graceful fallback when `buildOrderedSetInplace` returns null is untested — https://github.com/ClickHouse/ClickHouse/pull/103979

Could you review the test PR and add the `can be tested` label if it looks good?
Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**